ambry-router/src/main/java/com/github/ambry/router/OperationQuotaCharger.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2021 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.quota.Chargeable;
+import com.github.ambry.quota.QuotaChargeCallback;
+import com.github.ambry.quota.QuotaResource;
+import com.github.ambry.rest.RestServiceException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * {@link Chargeable} implementation for cases where quota is charged just once for entire operation.
+ */
+public class OperationQuotaCharger implements Chargeable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(OperationQuotaCharger.class);
+  private final QuotaChargeCallback quotaChargeCallback;
+  private final BlobId blobId;
+  private final String operationName;
+  private boolean isCharged;
+
+  /**
+   * Constructor for {@link OperationQuotaCharger}.
+   *
+   * @param quotaChargeCallback {@link QuotaChargeCallback} object to charge and check quotas.
+   * @param blobId {@link BlobId} of the blob for which quota will be charged.
+   * @param operationName Name of the operation.
+   */
+  public OperationQuotaCharger(QuotaChargeCallback quotaChargeCallback, BlobId blobId, String operationName) {
+    this.quotaChargeCallback = quotaChargeCallback;
+    this.blobId = blobId;
+    this.operationName = operationName;
+    this.isCharged = false;
+  }
+
+  @Override
+  public boolean check() {
+    if(quotaChargeCallback == null || isCharged) {
+      return true;
+    }
+    return quotaChargeCallback.check();
+  }
+
+  @Override
+  public boolean charge() {
```

Comment:
The current implementation is:

Create `QuotaChargeCallback` when submit a request to router. The `QuotaChargeCallback` is cost + action. When the code arrives `OperationQuotaCharger`, action is taken. 

I am thinking if we can simplify this to:
Only build `cost info` when submit a request. When the code arrives `OperationQuotaCharger`, it takes in `cost info` and charge(action).   All the actions(charge, check, etc) are inside OperationQuotaCharger.

Reply:
As discussed offline, since such a change spans multiple PR, we can take this up as a cleanup work after merging this.